tests/api/views/test_tool_deployment.py

```diff
@@ -1,29 +1,32 @@
 from unittest.mock import patch
 
+from model_mommy import mommy
 import pytest
 from rest_framework import status
 from rest_framework.reverse import reverse
 
-from controlpanel.api.tools import Tool
-
 
 @pytest.yield_fixture(autouse=True)
-def tool_deploy():
-    with patch.object(Tool, 'deploy_for') as deploy:
+def deploy_tool():
+    with patch('controlpanel.api.models.tool.cluster.deploy_tool') as deploy:
         yield deploy
 
 
-def test_create_when_invalid_tool_name(client):
+@pytest.fixture
+def tool(db):
+    return mommy.make('api.Tool')
+
+
+def test_create_when_invalid_tool_name(client, users):
+    client.force_login(users['normal_user'])
     tool_name = 'unsupported-tool'
     response = client.post(reverse('deployment-list'), {"name": tool_name})
     assert response.status_code == status.HTTP_400_BAD_REQUEST
 
 
-def test_create_when_valid_tool_name(client, tool_deploy, users):
-    tool_name = 'rstudio'
-    user = users['normal_user']
-    client.force_login(user)
-    response = client.post(reverse('deployment-list'), {"name": tool_name})
+def test_create_when_valid_tool_name(client, deploy_tool, tool, users):
```

Comment:
Probably stupid question: Who is calling this method and passing the `tool`/`deploy_tool`/etc arguments?

Reply:
You should read the [pytest documentation](https://docs.pytest.org/en/latest/).
Pytest finds all functions that match a specified pattern (by default `test_*`), executes them and passes in the fixtures that match the argument names.
* `client` is a fixture defined in `pytest-django` that returns the Django test client
* `deploy_tool` is defined above and returns a mock patch of the `deploy_tool` function in `controlpanel.api.cluster`
* `tool` is also defined above and returns a model instance
* `users` is defined in `tests/conftest.py`. (pytest evaluates all the `conftest.py` files in the path to the current test function)

Pytest seems a little magical to start with, but it is the best and I think most popular Python test framework.

Reply:
> users is defined in tests/conftest.py. (pytest evaluates all the conftest.py files in the path to the current test function)

😮 now I see where that comes from. VS Code seems confused by pytest's magical fixtures but I can see the logic now.

Regarding the two fixtures defined above the tests... 🤦‍♂ my brain must have been fried yesterday afternoon 😆